File: Java/1209. Remove All Adjacent Duplicates in String II.java
import java.util.*;
// import javafx.util.Pair;

class Solution {
    public String removeDuplicates(String s, int k) {
        ArrayDeque<Pair<Character, Integer>> stack = new ArrayDeque<Pair<Character, Integer>>();
        int count = 1;
        for(int i = 0, n = s.length() ; i < n ; i++) { 
            char c = s.charAt(i);
            count = (stack.isEmpty() || stack.getFirst().getKey()!=c)? 1 : stack.getFirst().getValue()+1;
            stack.addFirst(new Pair<Character, Integer>(c, count));
            if (stack.getFirst().getValue() >= k){
                for (int j = 0; j < k; j++) stack.removeFirst();
            }
        }
        StringBuilder res = new StringBuilder();
        while (!stack.isEmpty()){
            res.append(stack.removeFirst().getKey());
        }
        res.reverse();
        return res.toString();
    }
}

File: Kotlin/1209. Remove All Adjacent Duplicates in String II.kt
import java.util.ArrayDeque

class Solution {
    fun removeDuplicates(s: String, k: Int): String {
        var stack = ArrayDeque<Pair<Char, Int>>()
        var count = 1
        for (ch in s){
            count = if (stack.isEmpty() || stack.peek().first!=ch) 1 else stack.peek().second+1
            stack.push(Pair(ch, count))
            if (stack.peek().second >= k){
                for (i in 1..k) stack.pop()
            }
        }
        var res = "";
        for (ele in stack) res += ele.first
        return res.reversed()
    }
}